// dc-extensions-sdk/dist/lib/lib/Events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var CONTEXT;
(function (CONTEXT) {
    CONTEXT["GET"] = "context-get";
})(CONTEXT = exports.CONTEXT || (exports.CONTEXT = {}));
var CONTENT_ITEM;
(function (CONTENT_ITEM) {
    CONTENT_ITEM["GET"] = "content-item-get";
})(CONTENT_ITEM = exports.CONTENT_ITEM || (exports.CONTENT_ITEM = {}));
var MEDIA_LINK;
(function (MEDIA_LINK) {
    MEDIA_LINK["IMAGE_GET"] = "media-image-get";
    MEDIA_LINK["VIDEO_GET"] = "media-video-get";
})(MEDIA_LINK = exports.MEDIA_LINK || (exports.MEDIA_LINK = {}));
var CONTENT_LINK;
(function (CONTENT_LINK) {
    CONTENT_LINK["CONTENT_GET"] = "content-link-get";
})(CONTENT_LINK = exports.CONTENT_LINK || (exports.CONTENT_LINK = {}));
var CONTENT_REFERENCE;
(function (CONTENT_REFERENCE) {
    CONTENT_REFERENCE["CONTENT_REF_GET"] = "content-reference-get";
})(CONTENT_REFERENCE = exports.CONTENT_REFERENCE || (exports.CONTENT_REFERENCE = {}));
var FIELD;
(function (FIELD) {
    FIELD["MODEL_GET"] = "field-model-get";
    FIELD["MODEL_SET"] = "field-model-set";
    FIELD["MODEL_RESET"] = "field-model-reset";
    FIELD["MODEL_IS_VALID"] = "field-model-is-valid";
    FIELD["MODEL_VALIDATE"] = "field-model-validate";
    FIELD["SCHEMA_GET"] = "field-schema-get";
})(FIELD = exports.FIELD || (exports.FIELD = {}));
var FORM;
(function (FORM) {
    FORM["READ_ONLY"] = "form-read-only-change";
    FORM["GET_FORM_MODEL"] = "form-model-get";
})(FORM = exports.FORM || (exports.FORM = {}));
var FRAME;
(function (FRAME) {
    FRAME["HEIGHT_GET"] = "height-get";
    FRAME["HEIGHT_SET"] = "height-set";
})(FRAME = exports.FRAME || (exports.FRAME = {}));
